### ISSUE TITLE
when a single cohort in a cohort set is empty I have issues with omopgenerics::cohortTable default cohort_attrition creation

### ISSUE THREAD

**Issue by ablack3** [darwin-eu-dev/omopgenerics#135]
I think that `omopgenerics:::defaultCohortAttrition` should produce a table with at least one row per number of rows in the cohort_set. So if a cohort has a record in cohort_set it should also have a record in the default cohort attrition table. Currently this is not the case as far as I can tell.

``` r
library(CDMConnector)
library(dplyr)

con <- DBI::dbConnect(duckdb::duckdb(), eunomia_dir())
cdm <- cdm_from_con(con, "main", "main")

cohort <- tibble::tribble(
  ~cohort_definition_id, ~subject_id, ~cohort_start_date, ~cohort_end_date,
  1L,         22L,       "2010-05-01",     "2010-12-31",
  1L,         30L,       "2008-10-10",     "2010-12-31",
  1L,         81L,       "2008-11-29",     "2010-12-31"
) %>% mutate_if(stringr::str_detect(names(.), "date"), as.Date)

cdm <- insertTable(cdm, "cohort", cohort)

cohort_set <- tibble::tribble(
~cohort_definition_id, ~cohort_name,  ~limit, ~prior_observation, ~future_observation,                          ~end,
                   1L,  "gibleed_1", "first",                  0,                   0, "observation_period_end_date",
                   2L,  "gibleed_2", "first",                  0,                   0, "observation_period_end_date"
)
  

attrition_df <- omopgenerics:::defaultCohortAttrition(cohort, cohort_set)
attrition_df
#> # A tibble: 1 × 7
#>   cohort_definition_id number_records number_subjects reason_id reason          
#>                  <int>          <dbl>           <dbl>     <dbl> <chr>           
#> 1                    1              3               3         1 Initial qualify…
#> # ℹ 2 more variables: excluded_records <dbl>, excluded_subjects <dbl>

# note there is only one row in the attrition table which causes an error here
omopgenerics::cohortTable(cdm$cohort, cohort_set, cohortAttritionRef = attrition_df)
#> Error in `validateGeneratedCohortSet()` at omopgenerics/R/classCohortTable.R:53:3:
#> ! Present cohort_definition_id must be the same:
#> • cohort_set: 1 and 2
#> • cohort_attrition: 1
#> Backtrace:
#>     ▆
#>  1. └─omopgenerics::cohortTable(cdm$cohort, cohort_set, cohortAttritionRef = attrition_df)
#>  2.   └─omopgenerics:::validateGeneratedCohortSet(cohort, soft = .softValidation) at omopgenerics/R/classCohortTable.R:53:3
#>  3.     └─cli::cli_abort(...) at omopgenerics/R/classCohortTable.R:144:5
#>  4.       └─rlang::abort(...)


cdmDisconnect(cdm)
```

<sup>Created on 2024-01-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by ablack3** [comment]
Instead of left join here I think we need a slightly different operation.
https://github.com/darwin-eu-dev/omopgenerics/blob/2ed0a1b98c12c2e82ee1297d4511a60451e5e6b0/R/classCohortTable.R#L253
Maybe left_join(cohort_set, cohort_counts) or a full join. anyway if the cohort table has no records we need a 0 in the attrition and counts tables.

**Comment by catalamarti** [comment]
this should be fixed here @ablack3 #138
